File: project_euler/Problem061/Problem061.go
package project_euler

import (
    "fmt"
    "math"
)

func Problem061() {
    target := 6
    polygonals := make([]Polygonal, 0)
    addTriangle(&polygonals)
    addSquare(&polygonals)
    addPentagonal(&polygonals)
    addHexagonal(&polygonals)
    addHeptagonal(&polygonals)
    addOctagonal(&polygonals)

    length := len(polygonals)
    edges := make([][]int, length)
    for i := 0; i < length; i++ {
        edges[i] = make([]int, length)
        for j := 0; j < length; j++ {
            edges[i][j] = 0
        }
    }
    for i := 0; i < length; i++ {
        pivot := polygonals[i]
        for j := i + 1; j < length; j++ {
            polygonal := polygonals[j]
            if pivot.step != polygonal.step {
                if pivot.tail == polygonal.head {
                    edges[i][j] = 1
                }
                if polygonal.tail == pivot.head {
                    edges[j][i] = 1
                }
            }
        }
    }

    indexes := make([]int, 0)
    steps := make([]bool, target)
    for i := 0; i < target; i++ {
        steps[i] = false
    }
    pivot := polygonals[0].step
    steps[pivot] = true
    for i := 0; i < length; i++ {
        indexes = append(indexes, i)
        step := polygonals[i].step
        if step != pivot {
            return
        }
        search(polygonals, edges, i, indexes, steps, target)
        indexes = indexes[:len(indexes) - 1]
    }
}

func addTriangle(polygonals *[]Polygonal) {
    start := int(math.Ceil((math.Sqrt(8001) - 1) / 2))
    end := int(math.Ceil((math.Sqrt(80001) - 1) / 2))
    for i := start; i < end; i++ {
        value := i * (i + 1) >> 1
        if value % 100 >= 10 {
            *polygonals = append(*polygonals, Polygonal{value / 100, value % 100, 0})
        }
    }
}

func addSquare(polygonals *[]Polygonal) {
    start := int(math.Ceil(math.Sqrt(1000)))
    end := int(math.Ceil(math.Sqrt(10000)))
    for i := start; i < end; i++ {
        value := i * i
        if value % 100 >= 10 {
            *polygonals = append(*polygonals, Polygonal{value / 100, value % 100, 1})
        }
    }
}

func addPentagonal(polygonals *[]Polygonal) {
    start := int(math.Ceil((math.Sqrt(24001) + 1) / 6))
    end := int(math.Ceil((math.Sqrt(240001) + 1) / 6))
    for i := start; i < end; i++ {
        value := i * (3 * i - 1) >> 1
        if value % 100 >= 10 {
            *polygonals = append(*polygonals, Polygonal{value / 100, value % 100, 2})
        }
    }
}

func addHexagonal(polygonals *[]Polygonal) {
    start := int(math.Ceil((math.Sqrt(8001) + 1) / 4))
    end := int(math.Ceil((math.Sqrt(80001) + 1) / 4))
    for i := start; i < end; i++ {
        value := i * ((i << 1) - 1)
        if value % 100 >= 10 {
            *polygonals = append(*polygonals, Polygonal{value / 100, value % 100, 3})
        }
    }
}

func addHeptagonal(polygonals *[]Polygonal) {
    start := int(math.Ceil((math.Sqrt(40009) + 3) / 10))
    end := int(math.Ceil((math.Sqrt(400009) + 3) / 10))
    for i := start; i < end; i++ {
        value := i * (5 * i - 3) >> 1
        if value % 100 >= 10 {
            *polygonals = append(*polygonals, Polygonal{value / 100, value % 100, 4})
        }
    }
}

func addOctagonal(polygonals *[]Polygonal) {
    start := int(math.Ceil((math.Sqrt(12002) + 2) / 6))
    end := int(math.Ceil((math.Sqrt(120002) + 2) / 6))
    for i := start; i < end; i++ {
        value := i * (3 * i - 2)
        if value % 100 >= 10 {
            *polygonals = append(*polygonals, Polygonal{value / 100, value % 100, 5})
        }
    }
}

func search(polygonals []Polygonal, edges [][]int, start int, indexes []int, steps []bool, target int) {
    if len(indexes) == target {
        if polygonals[indexes[0]].head == polygonals[indexes[target - 1]].tail {
            result := 0
            for _, index := range indexes {
                vertex := polygonals[index]
                result += vertex.head * 100 + vertex.tail
            }
            fmt.Println(result)
        }
        return
    }

    for i := len(polygonals) - 1; i >= 0; i-- {
        if 0 != edges[start][i] {
            step := polygonals[i].step
            if !steps[step] {
                indexes = append(indexes, i)
                steps[step] = true
                search(polygonals, edges, i, indexes, steps, target)
                indexes = indexes[:len(indexes) - 1]
                steps[step] = false
            }
        }
    }
}

type Polygonal struct {
    head int
    tail int
    step int
}

File: project_euler/Problem061/Problem061.java
package project_euler;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Problem061 {
    public void solution() {
        int target = 6;
        List<Polygonal> polygonals = new ArrayList<>();
        addTriangle(polygonals);
        addSquare(polygonals);
        addPentagonal(polygonals);
        addHexagonal(polygonals);
        addHeptagonal(polygonals);
        addOctagonal(polygonals);

        int length = polygonals.size();
        int[][] edges = new int[length][length];
        for (int i = 0; i < length; ++i) {
            Arrays.fill(edges[i], 0);
        }
        for (int i = 0; i < length; ++i) {
            Polygonal pivot = polygonals.get(i);
            for (int j = i + 1; j < length; ++j) {
                Polygonal polygonal = polygonals.get(j);
                if (pivot.step != polygonal.step) {
                    if (pivot.tail == polygonal.head) {
                        edges[i][j] = 1;
                    }
                    if (polygonal.tail == pivot.head) {
                        edges[j][i] = 1;
                    }
                }
            }
        }

        List<Integer> indexes = new ArrayList<>();
        boolean[] steps = new boolean[target];
        Arrays.fill(steps, false);
        int pivot = polygonals.get(0).step;
        steps[pivot] = true;
        for (int i = 0; i < length; ++i) {
            indexes.add(i);
            int step = polygonals.get(i).step;
            if (step != pivot) {
                return;
            }
            search(polygonals, edges, i, indexes, steps, target);
            indexes.remove(indexes.size() - 1);
        }
    }

    private void addTriangle(List<Polygonal> polygonals) {
        int start = (int) Math.ceil((Math.sqrt(8001) - 1) / 2);
        int end = (int) Math.ceil((Math.sqrt(80001) - 1) / 2);
        for (int i = start; i < end; ++i) {
            int value = i * (i + 1) >> 1;
            if (value % 100 >= 10) {
                polygonals.add(new Polygonal(value, 0));
            }
        }
    }

    private void addSquare(List<Polygonal> polygonals) {
        int start = (int) Math.ceil(Math.sqrt(1000));
        int end = (int) Math.ceil(Math.sqrt(10000));
        for (int i = start; i < end; ++i) {
            int value = i * i;
            if (value % 100 >= 10) {
                polygonals.add(new Polygonal(value, 1));
            }
        }
    }

    private void addPentagonal(List<Polygonal> polygonals) {
        int start = (int) Math.ceil((Math.sqrt(24001) + 1) / 6);
        int end = (int) Math.ceil((Math.sqrt(240001) + 1) / 6);
        for (int i = start; i < end; ++i) {
            int value = i * (3 * i - 1) >> 1;
            if (value % 100 >= 10) {
                polygonals.add(new Polygonal(value, 2));
            }
        }
    }

    private void addHexagonal(List<Polygonal> polygonals) {
        int start = (int) Math.ceil((Math.sqrt(8001) + 1) / 4);
        int end = (int) Math.ceil((Math.sqrt(80001) + 1) / 4);
        for (int i = start; i < end; ++i) {
            int value = i * ((i << 1) - 1);
            if (value % 100 >= 10) {
                polygonals.add(new Polygonal(value, 3));
            }
        }
    }

    private void addHeptagonal(List<Polygonal> polygonals) {
        int start = (int) Math.ceil((Math.sqrt(40009) + 3) / 10);
        int end = (int) Math.ceil((Math.sqrt(400009) + 3) / 10);
        for (int i = start; i < end; ++i) {
            int value = i * (5 * i - 3) >> 1;
            if (value % 100 >= 10) {
                polygonals.add(new Polygonal(value, 4));
            }
        }
    }

    private void addOctagonal(List<Polygonal> polygonals) {
        int start = (int) Math.ceil((Math.sqrt(12002) + 2) / 6);
        int end = (int) Math.ceil((Math.sqrt(120002) + 2) / 6);
        for (int i = start; i < end; ++i) {
            int value = i * (3 * i - 2);
            if (value % 100 >= 10) {
                polygonals.add(new Polygonal(value, 5));
            }
        }
    }

    private void search(List<Polygonal> polygonals, int[][] edges, int start, List<Integer> indexes, boolean[] steps, int target) {
        if (indexes.size() == target) {
            if (polygonals.get(indexes.get(0)).head == polygonals.get(indexes.get(target - 1)).tail) {
                int result = 0;
                for (int index : indexes) {
                    Polygonal polygonal = polygonals.get(index);
                    result += polygonal.head * 100 + polygonal.tail;
                }
                System.out.println(result);
            }
            return;
        }

        for (int i = polygonals.size() - 1; i >= 0; --i) {
            if (0 != edges[start][i]) {
                int step = polygonals.get(i).step;
                if (!steps[step]) {
                    indexes.add(i);
                    steps[step] = true;
                    search(polygonals, edges, i, indexes, steps, target);
                    indexes.remove(indexes.size() - 1);
                    steps[step] = false;
                }
            }
        }
    }

    private static class Polygonal {
        public int head;
        public int tail;
        public int step;

        public Polygonal(int value, int step) {
            this.head = value / 100;
            this.tail = value % 100;
            this.step = step;
        }
    }
}
